File: elixir/test/station_test.exs
# RUN: make test
# RUN-SOME: mix test test/station_test.exs
#
# Focused unit tests for the contracts the JSON corpus cannot express
# (design station.md 13): the ambient instance, the event ring + tap,
# the env-only broker's hit/miss/refusal semantics, and the profile
# lookup walk. The struct-facing adapter half is exercised by the
# generated-SDK integration flow, not here - no SDK (and so no
# Voxgig.Struct) is present in a standalone library checkout.

defmodule Voxgig.StationTest do
  use ExUnit.Case, async: false

  alias Voxgig.Station
  alias Voxgig.Station.Descriptor
  alias Voxgig.Station.Error
  alias Voxgig.Station.Events
  alias Voxgig.Station.Profile
  alias Voxgig.Station.Secrets

  setup do
    Station.reset()
    :ok
  end

  # --- ambient instance (design station.md 10.2) ---

  test "open is idempotent; conflicting options are an error" do
    assert nil == Station.current()

    st = Station.open(%{"config" => nil})
    assert st == Station.open(%{"config" => nil})
    assert st == Station.current()

    err =
      assert_raise Error, fn ->
        Station.open(%{"config" => nil, "profile" => "prod"})
      end

    assert "station_open_conflict" == err.code

    Station.reset()
    assert nil == Station.current()
  end

  test "new stays isolated from the ambient instance" do
    ambient = Station.open(%{"config" => nil})
    isolated = Station.new(%{"config" => nil})
    assert ambient != isolated
    assert ambient == Station.current()
  end

  test "close is idempotent and drops the ambient slot" do
    st = Station.open(%{"config" => nil})
    assert false == Station.closed?(st)
    Station.close(st)
    assert true == Station.closed?(st)
    Station.close(st)
    assert nil == Station.current()
  end

  # --- construction events ---

  test "auto proxy degrades to solo with one warning event" do
    st = Station.new(%{"config" => nil})
    [first | _] = Station.events(st)
    assert "station" == first["kind"]
    assert first["meta"]["warn"] =~ "running solo"
  end

  test "proxy off emits no degradation warning" do
    st = Station.new(%{"config" => nil, "proxy" => "off"})
    assert [] == Station.events(st)
  end

  test "env-only honesty: unsupported provider kinds are said at construction" do
    config = %{
      "station" => 1,
      "profiles" => %{
        "default" => %{
          "secrets" => %{"providers" => [%{"kind" => "env"}, %{"kind" => "hashicorp"}]}
        }
      }
    }

    st = Station.new(%{"config" => config, "proxy" => "off"})
    [ev] = Station.events(st)
    assert ev["meta"]["warn"] =~ "no elixir sekreto port"
    assert ev["meta"]["warn"] =~ "hashicorp"
  end

  test "close warns on profile plugin keys matching no registered plugin" do
    config = %{
      "station" => 1,
      "profiles" => %{"default" => %{"plugin" => %{"typod" => %{"base" => "http://x"}}}}
    }

    st = Station.new(%{"config" => config, "proxy" => "off"})
    Station.close(st)
    [ev] = Station.events(st)
    assert ev["meta"]["warn"] =~ "typod"
    assert ev["meta"]["warn"] =~ "matched no registered plugin"
  end

  # --- the event ring (design station.md 6) ---

  test "ring is bounded, drops oldest, and counts drops in status" do
    buf = Events.buffer(3)

    {buf, _} = Events.emit(buf, %{"n" => 1})
    {buf, _} = Events.emit(buf, %{"n" => 2})
    {buf, _} = Events.emit(buf, %{"n" => 3})
    {buf, _} = Events.emit(buf, %{"n" => 4})
    {buf, _} = Events.emit(buf, %{"n" => 5})

    assert [%{"n" => 3}, %{"n" => 4}, %{"n" => 5}] == Events.events(buf)
    assert %{"buffered" => 3, "dropped" => 2} == Events.status(buf)
  end

  test "tap sees events, unsubscribes cleanly, and a raising tap never fails emit" do
    st = Station.new(%{"config" => nil, "proxy" => "off"})
    me = self()

    untap_raise = Station.tap(st, fn _ev -> raise "tap boom" end)
    untap = Station.tap(st, fn ev -> send(me, {:ev, ev}) end)

    Station.emit(st, %{"t" => 1, "kind" => "station", "meta" => %{"warn" => "w1"}})
    assert_received {:ev, %{"meta" => %{"warn" => "w1"}}}

    untap.()
    untap_raise.()
    Station.emit(st, %{"t" => 2, "kind" => "station", "meta" => %{"warn" => "w2"}})
    refute_received {:ev, %{"meta" => %{"warn" => "w2"}}}

    assert 2 == Station.status(st)["events"]["buffered"]
  end

  test "status shape" do
    st = Station.new(%{"config" => nil, "proxy" => "off"})
    status = Station.status(st)
    assert "solo" == status["mode"]
    assert "default" == status["profile"]
    assert [] == status["plugins"]
    assert %{"buffered" => 0, "dropped" => 0} == status["events"]
  end

  # --- the env-only broker (design station.md 2.2, 5.2) ---

  test "env provider resolves the sekreto env key, and caches per plugin" do
    System.put_env("STATIONTEST_ONE_APIKEY", "v-one")

    broker = Secrets.broker()
    providers = [%{"kind" => "env"}]

    {:ok, "v-one", broker} =
      Secrets.value(broker, providers, "stationtest-one", "stationtest_one.apikey")

    # Cached: a changed environment is not re-read until refresh.
    System.put_env("STATIONTEST_ONE_APIKEY", "v-two")

    {:ok, "v-one", broker} =
      Secrets.value(broker, providers, "stationtest-one", "stationtest_one.apikey")

    broker = Secrets.refresh(broker)

    {:ok, "v-two", _broker} =
      Secrets.value(broker, providers, "stationtest-one", "stationtest_one.apikey")

    System.delete_env("STATIONTEST_ONE_APIKEY")
  end

  test "a miss is not an error: station_secret_no_value when every store missed" do
    System.delete_env("STATIONTEST_MISSING_APIKEY")

    {:error, err} =
      Secrets.value(Secrets.broker(), [%{"kind" => "env"}], "p", "stationtest_missing.apikey")

    assert "station_secret_no_value" == err.code
    assert Exception.message(err) =~ "stationtest_missing.apikey"
  end

  test "a store this port cannot answer from is an error, never a fall-through" do
    # env misses first, then the chain reaches dotenv - which must stop
    # the chain (station.md 5.2), not silently skip to nothing.
    System.delete_env("STATIONTEST_CHAIN_APIKEY")

    {:error, err} =
      Secrets.value(
        Secrets.broker(),
        [%{"kind" => "env"}, %{"kind" => "dotenv", "file" => ".env"}],
        "p",
        "stationtest_chain.apikey"
      )

    assert "station_secret_error" == err.code
    assert Exception.message(err) =~ "env-only"
  end

  test "an env provider earlier in the chain answering never reaches the refusal" do
    System.put_env("STATIONTEST_FIRST_APIKEY", "hit")

    {:ok, "hit", _} =
      Secrets.value(
        Secrets.broker(),
        [%{"kind" => "env"}, %{"kind" => "hashicorp"}],
        "p",
        "stationtest_first.apikey"
      )

    System.delete_env("STATIONTEST_FIRST_APIKEY")
  end

  test "an invalid secret name is a station_secret_error" do
    {:error, err} = Secrets.value(Secrets.broker(), [%{"kind" => "env"}], "p", "Not A Name")
    assert "station_secret_error" == err.code
  end

  test "hoisted values win over the chain" do
    broker = Secrets.hoist(Secrets.broker(), "solardemo", "resident-key")
    {:ok, "resident-key", _} = Secrets.value(broker, [%{"kind" => "env"}], "solardemo", "x.apikey")
  end

  test "scrub is exact-value with no length floor" do
    broker = Secrets.broker()
    broker = Secrets.hoist(broker, "a", "k")
    broker = Secrets.hoist(broker, "b", "long-secret-value")

    out = Secrets.scrub(broker, "got k and long-secret-value back")
    refute out =~ "long-secret-value"
    assert out =~ "[redacted]"
    refute String.contains?(out, " k ")
  end

  test "redact runs the broker scrub through the station" do
    st = Station.new(%{"config" => nil, "proxy" => "off"})
    Station.hoist(st, "solardemo", "sekrit")
    assert false == String.contains?(Station.redact(st, "a sekrit b"), "sekrit")
    # The hoist itself is a visible warning event, never a silent downgrade.
    assert Enum.any?(Station.events(st), fn ev ->
             "station" == ev["kind"] and (ev["meta"]["warn"] || "") =~ "hoisted"
           end)
  end

  # --- profile lookup (design station.md 3.5) ---

  test "find_config_file walks cwd upward" do
    base = Path.join(System.tmp_dir!(), "station-test-#{System.unique_integer([:positive])}")
    nested = Path.join([base, "a", "b"])
    File.mkdir_p!(nested)
    File.write!(Path.join(base, "station.json"), ~s({ "station": 1 }))

    assert Path.join(base, "station.json") == Profile.find_config_file(nested)

    config = Profile.load_config(nested)
    assert %{"station" => 1} == config

    File.rm_rf!(base)
  end

  test "profile selection: option beats VOXGIG_STATION_PROFILE beats default" do
    System.delete_env("VOXGIG_STATION_PROFILE")
    assert "default" == Profile.select(nil)

    System.put_env("VOXGIG_STATION_PROFILE", "stage")
    assert "stage" == Profile.select(nil)
    assert "prod" == Profile.select("prod")
    System.delete_env("VOXGIG_STATION_PROFILE")
  end

  # --- canonical serializer edges beyond the corpus ---

  test "canonical serializer escapes control characters JSON-style" do
    assert "\"a\\u0001b\\nc\"" == Descriptor.canonical_serialize("a" <> <<1>> <> "b\nc")
  end

  test "envtoken trims and collapses non-alphanumerics" do
    assert "GNARLY_PETS" == Descriptor.envtoken("gnarly-pets")
    assert "WEIRD_NAME_2" == Descriptor.envtoken("Weird--Name..2")
    assert "" == Descriptor.envtoken(nil)
  end

  test "normalize: auth-inactive config plans no credential but keeps the view" do
    {descriptor, warnings} =
      Descriptor.normalize(%{
        "main" => %{"name" => "Solardemo", "slug" => "solardemo", "version" => "1.2.3",
          "target" => "elixir"},
        "options" => %{"base" => "http://localhost:8901"},
        "entity" => %{},
        "feature" => %{}
      })

    assert [] == warnings
    assert false == descriptor["auth"]["active"]
    assert "solardemo.apikey" == descriptor["auth"]["secretname"]
    assert "elixir" == descriptor["target"]
  end
end

File: swift/Tests/VoxgigStationTests/StationTest.swift
// RUN: make test   (or: swift test)
//
// Focused unit cases: the station mechanics the corpus cannot reach
// without an SDK - the event ring, the env-only broker's miss-vs-error
// and floor-less scrub, the binding guards, and the transport
// middleware's injection plan (the copy-on-inject CLONE itself is the
// generated adapter's one physical duty, exercised in the consumer
// end-to-end flow; what these cases pin is that the caller's header map
// is never mutated and only the sent set carries the value).

import Foundation
import VoxgigStation
import XCTest

#if canImport(Glibc)
  import Glibc
#endif

// A station over an in-memory profile, isolated from the ambient
// instance and from any station.json on disk. Secrets ride the process
// environment - the env provider is the only one this port has (station
// design 2.2).
func memStation(_ plugin: SJson? = nil, proxy: String? = nil) throws -> Station {
  var profile: [String: SJson] = [
    "secrets": .map(["providers": .list([.map(["kind": .str("env")])])])
  ]
  if let plugin = plugin {
    profile["sdk"] = plugin
  }
  var opts: [String: SJson] = [
    "config": .map([
      "station": .num(1),
      "profiles": .map(["default": .map(profile)]),
    ])
  ]
  if let proxy = proxy {
    opts["proxy"] = .str(proxy)
  }
  return try Station(.map(opts))
}

// The embedded-config shape a generated SDK carries, small.
func sdkConfig(_ name: String, _ slug: String?) -> SJson {
  var main: [String: SJson] = ["name": .str(name)]
  if let slug = slug {
    main["slug"] = .str(slug)
    main["version"] = .str("0.0.1")
    main["target"] = .str("swift")
  }
  return .map([
    "main": .map(main),
    "feature": .map(["test": .map([:])]),
    "options": .map([
      "base": .str("http://localhost:8000"),
      "auth": .map(["prefix": .str("")]),
      "entity": .map(["todo": .map([:])]),
    ]),
    "entity": .map([
      "todo": .map([
        "name": .str("todo"),
        "fields": .list([.map(["name": .str("title"), "kind": .str("String")])]),
        "op": .map([
          "list": .map([
            "points": .list([
              .map([
                "method": .str("GET"),
                "orig": .str("/api/todo"),
                "parts": .list([.str("api"), .str("todo")]),
              ])
            ])
          ])
        ]),
      ])
    ]),
  ])
}

@discardableResult
func bind(
  _ st: Station, _ client: AnyObject,
  _ featureOpts: SJson = .map([:]),
  _ residentApikey: String = "",
  _ featureNames: [String] = ["station"]
) throws -> Bound? {
  return try st.featureBinding(
    client, featureNames, sdkConfig("TestPlug", "test-plug"),
    .map([:]), featureOpts, residentApikey)
}

func hasWarn(_ st: Station, _ needle: String) -> Bool {
  for ev in st.events() {
    if let warn = ev.get("meta").get("warn").asStr, warn.contains(needle) {
      return true
    }
  }
  return false
}

final class StationTest: XCTestCase {

  override func setUp() {
    super.setUp()
    Station.reset()
    unsetenv("TEST_PLUG_APIKEY")
  }

  override func tearDown() {
    Station.reset()
    unsetenv("TEST_PLUG_APIKEY")
    super.tearDown()
  }

  func testEventsRingOverflowDropsOldestCounted() {
    let buffer = EventBuffer(2)
    buffer.emit(.map(["t": .num(1), "kind": .str("station")]))
    buffer.emit(.map(["t": .num(2), "kind": .str("station")]))
    buffer.emit(.map(["t": .num(3), "kind": .str("station")]))
    let events = buffer.events()
    XCTAssertEqual(2, events.count, "ring bounded")
    XCTAssertEqual(2, events[0].get("t").asNum, "oldest dropped")
    XCTAssertEqual(1, buffer.status().get("dropped").asNum, "drop counted")
  }

  func testEventsTapUnsubscribes() {
    let buffer = EventBuffer()
    var seen = 0
    let off = buffer.tap { _ in seen += 1 }
    buffer.emit(.map(["kind": .str("station")]))
    XCTAssertEqual(1, seen, "tap saw the event")
    off()
    buffer.emit(.map(["kind": .str("station")]))
    XCTAssertEqual(1, seen, "unsubscribed")
  }

  func testBrokerEnvResolveMissAndFloorlessScrub() throws {
    setenv("A_APIKEY", "xy", 1)
    defer { unsetenv("A_APIKEY") }

    let broker = SecretBroker(.list([.map(["kind": .str("env")])]))
    XCTAssertEqual("xy", try broker.value("a", "a.apikey"), "resolves from env")
    // Exact-value scrub, even below sekreto's four-character floor.
    XCTAssertEqual("k=[redacted];", broker.scrub("k=xy;"), "floor-less scrub")

    do {
      _ = try broker.value("b", "b.apikey")
      XCTFail("miss should throw")
    } catch let err as StationError {
      XCTAssertEqual("station_secret_no_value", err.code, "miss code")
    }
  }

  func testBrokerRefusesNonEnvProviderKinds() {
    // Env-only honesty (design 2.2/5.2): a store this port cannot answer
    // from is an ERROR, never a fall-through to a weaker store.
    let broker = SecretBroker(
      .list([
        .map(["kind": .str("hashicorp"), "addr": .str("https://v")]),
        .map(["kind": .str("env")]),
      ]))
    do {
      _ = try broker.value("a", "a.apikey")
      XCTFail("non-env provider must refuse")
    } catch let err as StationError {
      XCTAssertEqual("station_secret_error", err.code, "refusal code")
      XCTAssertTrue(err.message.contains("no swift sekreto port"), "says so")
    } catch {
      XCTFail("unexpected error type")
    }
    XCTAssertEqual(["hashicorp"], broker.unsupportedKinds(), "kinds listed")
  }

  func testBrokerRefreshDropsCacheKeepsHoisted() throws {
    setenv("A_APIKEY", "one", 1)
    defer { unsetenv("A_APIKEY") }

    let broker = SecretBroker(.null)
    XCTAssertEqual("one", try broker.value("a", "a.apikey"))
    setenv("A_APIKEY", "two", 1)
    XCTAssertEqual("one", try broker.value("a", "a.apikey"), "cached")
    broker.refresh()
    XCTAssertEqual("two", try broker.value("a", "a.apikey"), "re-resolved")

    broker.hoist("h", "hoisted")
    broker.refresh()
    XCTAssertEqual("hoisted", try broker.value("h", "h.apikey"), "override survives")
  }

  func testOpenIdempotentAmbientConflictingReopenFails() throws {
    Station.reset()
    defer { Station.reset() }

    let one = try Station.open(.map(["config": .null]))
    XCTAssertTrue(one === (try Station.open(.map(["config": .null]))), "same instance")
    XCTAssertTrue(one === Station.current(), "current() is the ambient")

    do {
      _ = try Station.open(.map(["config": .null, "profile": .str("prod")]))
      XCTFail("conflicting open must fail")
    } catch let err as StationError {
      XCTAssertEqual("station_open_conflict", err.code, "conflict code")
    }

    one.close()
    XCTAssertNil(Station.current(), "close() of the ambient resets it")
  }

  func testOptionsActivationEntryCarriesHandle() throws {
    let st = try memStation()
    let options = st.options(["base": "http://x:1"])
    let fmap = options["feature"] as? [String: Any?]
    let entry = fmap?["station"] as? [String: Any?]
    XCTAssertEqual(true, entry?["active"] as? Bool, "active")
    XCTAssertTrue(st === (entry?["station"] as? Station), "handle rides the options")
    XCTAssertEqual("http://x:1", options["base"] as? String, "caller opts kept")
    XCTAssertNotNil(entry?["calleropts"] as? [String: Any?], "calleropts marker")
  }

  func testBindingRegistersPlantsPlaceholderHoistsResident() throws {
    setenv("TEST_PLUG_APIKEY", "real-key-1", 1)
    let st = try memStation()
    let client = NSObject()

    let bound = try bind(st, client, .map(["calleropts": .map([:])]), "resident-key")
    XCTAssertEqual("test-plug", bound?.binding.slug, "slug")
    XCTAssertEqual("[station:test-plug]", bound?.placeholder, "placeholder planned")
    XCTAssertTrue(hasWarn(st, "hoisted"), "hoist warning emitted")
    XCTAssertEqual("x [redacted] y", st.redact("x resident-key y"), "hoisted value scrubbed")

    var sawConstruct = false
    for ev in st.events() where "construct" == ev.get("kind").asStr {
      sawConstruct = true
      XCTAssertEqual("test-plug", ev.get("plugin").asStr)
      XCTAssertEqual("R1", ev.get("meta").get("rung").asStr)
    }
    XCTAssertTrue(sawConstruct, "construct event emitted")

    // Same client, second arrival: inert.
    XCTAssertNil(try bind(st, client), "idempotent per client")

    // A second client of the same SDK: refused.
    do {
      _ = try bind(st, NSObject())
      XCTFail("second client must fail")
    } catch let err as StationError {
      XCTAssertEqual("station_bound_twice", err.code, "bound twice code")
    }
  }

  func testBindingWrongWrapPositionFailsLoudly() throws {
    let st = try memStation()
    do {
      _ = try bind(st, NSObject(), .map([:]), "", ["test", "retry", "station"])
      XCTFail("order guard must trip")
    } catch let err as StationError {
      XCTAssertEqual("station_wrap_order", err.code, "order code")
    }
  }

  func testBindingInertBaseEntriesAreSkippedByGuard() throws {
    // The generated makeFeature's default case absorbs an unknown
    // activation name as an inert BaseFeature (name "base") - it
    // occupies a features slot but never wraps the transport, so the
    // position guard must tolerate it wherever it sits.
    let st = try memStation()
    let bound = try bind(st, NSObject(), .map([:]), "", ["test", "base", "station"])
    XCTAssertNotNil(bound, "stray base between test and station tolerated")
  }

  func testBindingProfileBaseOnlyWithoutCallerBase() throws {
    let st = try memStation(.map(["test-plug": .map(["base": .str("http://profile:9")])]))
    let bound = try bind(st, NSObject(), .map(["calleropts": .map([:])]))
    XCTAssertEqual("http://profile:9", bound?.base, "profile base handed back")

    let st2 = try memStation(.map(["test-plug": .map(["base": .str("http://profile:9")])]))
    let bound2 = try bind(
      st2, NSObject(), .map(["calleropts": .map(["base": .str("http://caller:1")])]))
    XCTAssertNil(bound2?.base, "caller base wins")
  }

  func testTransportInjectsOnlyTheSentSet() throws {
    setenv("TEST_PLUG_APIKEY", "real-key-2", 1)
    let st = try memStation()
    let bound = try bind(st, NSObject())
    let binding = bound!.binding

    binding.opStart("ctx1")

    let headers = ["authorization": "[station:test-plug]"]
    var sent: String? = nil
    let res: [String: Any] = try binding.transport(
      "ctx1", true, "http://localhost:8000/api/todo", "GET", headers,
      { injected, _ in
        sent = injected?["authorization"]
        return ["status": 200]
      },
      { _ in (200, 12) })

    XCTAssertEqual("real-key-2", sent, "real value on the wire")
    XCTAssertEqual("[station:test-plug]", headers["authorization"],
      "caller's header map keeps the placeholder")
    XCTAssertEqual(200, res["status"] as? Int, "response returned")

    var http: SJson? = nil
    for ev in st.events() where "http" == ev.get("kind").asStr {
      http = ev
    }
    XCTAssertNotNil(http, "http event emitted")
    XCTAssertEqual(binding.corrOf("ctx1"), http?.get("corr").asStr, "correlated")
    XCTAssertEqual(200, http?.get("http").get("status").asNum, "wire status")
    XCTAssertEqual(12, http?.get("http").get("bytes").asNum, "bytes")
    XCTAssertEqual("localhost:8000", http?.get("http").get("host").asStr, "host")
    XCTAssertEqual("/api/todo", http?.get("http").get("path").asStr, "path")

    binding.opDone("ctx1", "todo", "list", "ok")
    var op: SJson? = nil
    for ev in st.events() where "op" == ev.get("kind").asStr {
      op = ev
    }
    XCTAssertEqual("ok", op?.get("op").get("outcome").asStr, "op event outcome")
    XCTAssertNotNil(op?.get("corr").asStr, "op event correlated")
    XCTAssertNil(binding.corrOf("ctx1"), "correlation closed")
  }

  func testTransportNoInjectionIntoMocks() throws {
    setenv("TEST_PLUG_APIKEY", "real-key-3", 1)
    let st = try memStation()
    let binding = (try bind(st, NSObject()))!.binding

    var sawInjected = false
    _ = try binding.transport(
      nil, false, "http://localhost:8000/api/todo", "GET",
      ["authorization": "[station:test-plug]"],
      { injected, _ in
        sawInjected = nil != injected
        return true
      },
      { _ in (200, 0) })
    XCTAssertFalse(sawInjected, "placeholder rides through untouched in mock mode")
  }

  func testTransportMissingSecretFailsTheOpWithCode() throws {
    let st = try memStation()
    let binding = (try bind(st, NSObject()))!.binding

    do {
      _ = try binding.transport(
        nil, true, "http://localhost:8000/api/todo", "GET", [:],
        { _, _ in true }, { _ in (200, 0) })
      XCTFail("missing secret must throw")
    } catch let err as StationError {
      XCTAssertEqual("station_secret_no_value", err.code, "miss code")
    }

    var recorded = false
    for ev in st.events() where "error" == ev.get("kind").asStr {
      recorded = recorded
        || "station_secret_no_value" == ev.get("err").get("code").asStr
    }
    XCTAssertTrue(recorded, "error event recorded")
  }

  func testTransportHostsPolicyDeniesOffListLiveOnly() throws {
    setenv("TEST_PLUG_APIKEY", "k", 1)
    let st = try memStation(
      .map([
        "test-plug": .map([
          "policy": .map(["hosts": .list([.str("api.good.example")])])
        ])
      ]))
    let binding = (try bind(st, NSObject()))!.binding

    do {
      _ = try binding.transport(
        nil, true, "http://evil.example/x", "GET", [:],
        { _, _ in true }, { _ in (200, 0) })
      XCTFail("off-list host must be denied")
    } catch let err as StationError {
      XCTAssertEqual("station_host_allow", err.code, "host code")
    }

    // A mock-transport call is not egress; the policy does not fire.
    let ok: Bool = try binding.transport(
      nil, false, "http://evil.example/x", "GET", [:],
      { _, _ in true }, { _ in (200, 0) })
    XCTAssertTrue(ok, "mock path unaffected")
  }

  func testTransportManualRedirectUnderHostsPolicy() throws {
    setenv("TEST_PLUG_APIKEY", "k", 1)
    let st = try memStation(
      .map([
        "test-plug": .map([
          "policy": .map(["hosts": .list([.str("api.good.example")])])
        ])
      ]))
    let binding = (try bind(st, NSObject()))!.binding

    var manual = false
    _ = try binding.transport(
      nil, true, "http://api.good.example/x", "GET", [:],
      { _, manualRedirect in
        manual = manualRedirect
        return true
      },
      { _ in (200, 0) })
    XCTAssertTrue(manual, "redirect: manual instructed under a hosts policy")
  }

  func testTransportRequireProxyFailsClosedOnOperationPath() throws {
    let st = try memStation(nil, proxy: "require")
    let binding = (try bind(st, NSObject()))!.binding

    do {
      _ = try binding.transport(
        nil, true, "http://localhost:1/x", "GET", [:],
        { _, _ in true }, { _ in (200, 0) })
      XCTFail("require must fail traffic")
    } catch let err as StationError {
      XCTAssertEqual("station_no_proxy", err.code, "no proxy code")
    }
  }

  func testEnvOnlyChainIsSaidAtConstruction() throws {
    let st = try Station(
      .map([
        "config": .map([
          "station": .num(1),
          "profiles": .map([
            "default": .map([
              "secrets": .map([
                "providers": .list([
                  .map(["kind": .str("hashicorp"), "addr": .str("https://v")])
                ])
              ])
            ])
          ]),
        ])
      ]))
    XCTAssertTrue(hasWarn(st, "no swift sekreto port"), "env-only chain said at construction")
  }

  func testCloseWarnsUnmatchedProfilePluginKeys() throws {
    let st = try memStation(.map(["typoed": .map(["base": .str("http://x")])]))
    st.close()
    XCTAssertTrue(hasWarn(st, "matched no registered plugin"), "typo'd key warned")
  }

  func testDescriptorSurface() throws {
    let st = try memStation()
    _ = try bind(st, NSObject())

    let descriptor = try st.descriptorOf("test-plug")
    XCTAssertEqual("TEST_PLUG", descriptor.get("envtoken").asStr)
    XCTAssertTrue(
      (try st.canonicalDescriptor("test-plug")).contains("\"slug\":\"test-plug\""))

    XCTAssertEqual(1, st.plugins().count, "registry queryable")
    XCTAssertEqual("solo", st.status().get("mode").asStr, "solo mode")

    do {
      _ = try st.descriptorOf("nope")
      XCTFail("unknown plugin must fail")
    } catch let err as StationError {
      XCTAssertEqual("station_no_plugin", err.code)
      XCTAssertTrue(err.message.contains("test-plug"), "candidates listed")
    }
  }
}
